pkg/ocm_upgrade_config_manager/ocm_upgrade_config_manager.go

```diff
@@ -0,0 +1,438 @@
+	req.Header.Add("Authorization", authVal)
+	transport := http.Transport{
+		TLSHandshakeTimeout: time.Second * 5,
+	}
+	return transport.RoundTrip(req)
+}
+
+// UpgradeConfigManager will trigger X every `WatchInterval` and only stop if the operator is killed or a
+// message is sent on the stopCh
+func (s *osdUpgradeConfigManager) Start(stopCh <-chan struct{}) {
+	log.Info("Starting the upgradeConfigManager")
+
+	initialSync := true
+	for {
+
+		// Select a new watch interval with jitter
+		duration := durationWithJitter(s.config.GetWatchInterval(), JITTER_FACTOR)
+
+		// Don't wait if this is the first run since starting
+		if initialSync {
+			duration = 0
+			initialSync = false
+		}
+
+		select {
+		case <-time.After(duration):
+			_, err := s.RefreshUpgradeConfig()
+			if err != nil {
+				log.Error(err, "unable to refresh upgrade config")
+			}
+		case <-stopCh:
+			log.Info("Stopping the upgradeConfigManager")
+			break
+		}
+	}
+}
+
+func readConfigManagerConfig(client client.Client) (*ocmUpgradeConfigManagerConfig, error) {
+	ns, err := getOperatorNamespace()
+	if err != nil {
+		return nil, err
+	}
+
+	cfb := configmanager.NewBuilder()
+	cfm := cfb.New(client, ns)
+	cfg := &ocmUpgradeConfigManagerConfig{}
+	err = cfm.Into(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+// The primary function called each watch interval to synchronise the cluster's
+// UpgradeConfig against the Cluster Service's Upgrade Policy.
+// Returns an indication of if the upgrade config had changed during the refresh,
+// and indication of error if one occurs.
+func (s *osdUpgradeConfigManager) RefreshUpgradeConfig() (bool, error) {
+
+	cluster, err := getClusterFromOCMApi(s.client, s.httpClient, s.config)
+	if err != nil {
+		log.Error(err, "cannot obtain internal cluster ID")
+		return false, err
+	}
+	if cluster.Id == "" {
+		return false, fmt.Errorf("cluster ID not found via OCM")
+	}
+	if cluster.Version.ChannelGroup == "" {
+		return false, fmt.Errorf("cluster channel group not set or found via OCM")
+	}
+
+	// Retrieve the cluster's upgrade policy from Cluster Services
+	upgradePolicies, err := getClusterUpgradePolicies(cluster, s.httpClient, s.config)
+	if err != nil {
+		log.Error(err, "cannot fetch next upgrade status")
+		return false, err
+	}
+
+	// Apply the Upgrade Policies to the cluster
+	changed, err := applyUpgradePolicies(upgradePolicies, cluster, s.client)
+	if err != nil {
+		log.Error(err, "cannot apply upgrade config")
+		return false, err
+	}
+
+	return changed, nil
+}
+
+// Queries and returns the Upgrade Policy from Cluster Services
+func getClusterUpgradePolicies(cluster *clusterInfo, client *http.Client, cfg *ocmUpgradeConfigManagerConfig) (*upgradePolicyList, error) {
+
+	url, err := url.Parse(cfg.ConfigManagerConfig.OcmBaseURL)
+	if err != nil {
+		log.Error(err, "OCM API URL unparsable.")
+		return nil, fmt.Errorf("can't parse OCM API url: %v", err)
+	}
```

Comment:
can use?
```
if cfg.IsValid != nil {
   return nil, fmt.Errorf("can't parse OCM API url: %v", err)
}
url := cfg.ConfigManagerConfig.OcmBaseURL
```

Reply:
I've fixed this up a bit in e77eca5. Enforced an IsValid() call when initially reading the config during creation of the Manager. Now in this part of the code it just reads the URL type from the config.